lib/components/room-upgrade-handler.js

```diff
@@ -0,0 +1,135 @@
+const log = require("./logging").get("RoomUpgradeHandler");
+const MatrixRoom = require("../models/rooms/matrix");
+
+/**
+ * Handles migration of rooms when a room upgrade is performed.
+ */
+class RoomUpgradeHandler {
+    /**
+     * @param {RoomUpgradeHandler~Options} opts
+     * @param {Bridge} bridge The parent bridge.
+     */
+    constructor(opts, bridge) {
+        this._opts = opts;
+        this._bridge = bridge;
+        this._waitingForInvite = new Map(); //newRoomId: oldRoomId
+    }
+
+    onTombstone(ev) {
+        const movingTo = ev.content.replacement_room;
+        log.info(`Got tombstone event for ${ev.room_id} -> ${movingTo}`);
+        // Try to join the new room.
+        return this._joinNewRoom(movingTo, movingTo).then(() => {
+            return this._onJoinedNewRoom(ev.room_id, movingTo);
+        }).catch((err) => {
+            log.error("Couldn't handle room upgrade: ", err);
+            // We can wait for an invite and try again.
+            this._waitingForInvite.set(movingTo, ev.room_id);
+        });
+
+    }
+
+    _joinNewRoom(newRoomId, roomIdOrAlias) {
+        const intent = this._bridge.getIntent();
+        return intent.join(roomIdOrAlias, true, true).catch((ex) => {
+            if (newRoomId !== roomIdOrAlias ||
+                !(ex.errcode == "M_UNKNONW" && ex.error === "No known servers")) {
+                // We need to wait to be invited
+                throw Error("Need to wait for invite");
+            }
+            // RoomId is not routable, so try to get the alias and join that.
```

Comment:
This is unsafe and shouldn't be done. Throw an error instead and let the appservice deal with it